contracts/Backing.sol

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/proxy/Initializable.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "./common/Scale.sol";
+import "./common/Ownable.sol";
+import { ScaleStruct } from "./common/Scale.struct.sol";
+import "./interfaces/IRelay.sol";
+import "./interfaces/IERC20Option.sol";
+
+contract Backing is Initializable, Ownable {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    struct BridgerInfo {
+        address target;
+        uint256 timestamp;
+    }
+
+    IRelay public relay;
+    bytes public substrateEventStorageKey;
+
+    mapping(address => BridgerInfo) public assets;
+    mapping(uint32 => address) public history;
+
+    struct Fee {
+        address token;
+        uint256 fee;
+    }
+    Fee public registerFee;
+    Fee public transferFee;
+
+    event NewTokenRegistered(address indexed token, string name, string symbol, uint8 decimals, uint256 fee);
+    event BackingLock(address indexed token, address target, uint256 amount, address receiver, uint256 fee);
+    event VerifyProof(uint32 blocknumber);
+    event RegistCompleted(address token, address target);
+    event RedeemTokenEvent(address token, address target, address receipt, uint256 amount);
+
+    function initialize(address _relay, address _registerFeeToken, address _transferFeeToken) public initializer {
+        ownableConstructor();
+        relay = IRelay(_relay);
+        registerFee = Fee(_registerFeeToken, 0);
+        transferFee = Fee(_transferFeeToken, 0);
+    }
+
+    function setStorageKey(bytes memory key) external onlyOwner {
+        substrateEventStorageKey = key;
+    }
+
+    function setRegisterFee(address token, uint256 fee) external onlyOwner {
+        registerFee.token = token;
+        registerFee.fee = fee;
+    }
+
+    function setTransferFee(address token, uint256 fee) external onlyOwner {
+        transferFee.token = token;
+        transferFee.fee = fee;
+    }
+
+    function registerToken(address token) external {
+        require(assets[token].timestamp == 0, "asset has been registered");
+        if (registerFee.fee > 0) {
+            IERC20Option(registerFee.token).burn(msg.sender, registerFee.fee);
```

Comment:
This looks not good because it directly burn token from msg.sender.

It might be better if  transferFrom msg.sender to this, and then burn from this.

Reply:
And some ERC20 token only allow to burn from this.